grovedb/src/replication.rs

```diff
@@ -383,277 +193,145 @@ impl GroveDb {
-            let restorer = Restorer::new(merk, app_hash, None);
-            root_prefix_state_sync_info.restorer = Some(restorer);
-            root_prefix_state_sync_info.pending_chunks.insert(vec![]);
-            state_sync_info
-                .current_prefixes
-                .insert(root_prefix, root_prefix_state_sync_info);
-            state_sync_info.app_hash = app_hash;
-        } else {
-            return Err(Error::InternalError(
-                "Unable to open merk for replication".to_string(),
-            ));
-        }
 
-        Ok(state_sync_info)
+        let mut session = self.start_syncing_session(app_hash);
+
+        session.add_subtree_sync_info(
+            self,
+            SubtreePath::empty(),
+            app_hash,
+            None,
+            root_prefix,
+            grove_version,
+        )?;
+
+        Ok(session)
     }
+}
 
-    // Apply a chunk (should be called by ABCI when ApplySnapshotChunk method is
-    // called) Params:
-    // state_sync_info: Consumed MultiStateSyncInfo
-    // global_chunk_id: Global chunk id
-    // chunk: Chunk proof operators encoded in bytes
-    // tx: Transaction for the state sync
-    // Returns the next set of global chunk ids that can be fetched from sources (+
-    // the MultiStateSyncInfo transferring ownership back to the caller)
-    pub fn apply_chunk<'db>(
-        &'db self,
-        mut state_sync_info: MultiStateSyncInfo<'db>,
-        global_chunk_id: &[u8],
-        chunk: Vec<u8>,
-        tx: &'db Transaction,
-        version: u16,
-        grove_version: &GroveVersion,
-    ) -> Result<(Vec<Vec<u8>>, MultiStateSyncInfo), Error> {
-        check_grovedb_v0!(
-            "apply_chunk",
-            grove_version.grovedb_versions.replication.apply_chunk
-        );
-        // For now, only CURRENT_STATE_SYNC_VERSION is supported
-        if version != CURRENT_STATE_SYNC_VERSION {
-            return Err(Error::CorruptedData(
-                "Unsupported state sync protocol version".to_string(),
-            ));
-        }
-        if version != state_sync_info.version {
-            return Err(Error::CorruptedData(
-                "Unsupported state sync protocol version".to_string(),
-            ));
-        }
+// Converts a path into a human-readable string (for debugging)
+pub fn util_path_to_string(path: &[Vec<u8>]) -> Vec<String> {
+    let mut subtree_path_str: Vec<String> = vec![];
+    for subtree in path {
+        let string = std::str::from_utf8(subtree).unwrap_or("<NON_UTF8_PATH>");
+        subtree_path_str.push(string.to_string());
+    }
+    subtree_path_str
+}
 
-        let mut next_chunk_ids = vec![];
+// Splits the given global chunk id into
+// [SUBTREE_PREFIX:SIZE_ROOT_KEY:ROOT_KEY:IS_SUM_TREE:CHUNK_ID]
+pub fn util_split_global_chunk_id(
+    global_chunk_id: &[u8],
+    app_hash: &[u8],
+) -> Result<(crate::SubtreePrefix, Option<Vec<u8>>, bool, Vec<u8>), Error> {
```

Comment:
Linter is right: instead of big unnamed tuple better to define a struct with named fields and use this functions as a constructor method of it